File: java/test/behaviour/concept/thing/ThingSteps.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.test.behaviour.concept.thing;

import com.vaticle.typedb.client.api.concept.thing.Thing;
import com.vaticle.typedb.client.api.concept.type.ThingType;
import com.vaticle.typedb.client.common.Label;
import com.vaticle.typedb.client.test.behaviour.config.Parameters.RootLabel;
import io.cucumber.java.After;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

import java.util.HashMap;
import java.util.Map;

import static com.vaticle.typedb.client.api.concept.type.ThingType.Annotation.key;
import static com.vaticle.typedb.client.test.behaviour.concept.type.thingtype.ThingTypeSteps.get_thing_type;
import static com.vaticle.typedb.client.test.behaviour.connection.ConnectionStepsBase.tx;
import static com.vaticle.typedb.client.test.behaviour.util.Util.assertThrows;
import static com.vaticle.typedb.common.collection.Collections.set;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

public class ThingSteps {
    private static Map<String, Thing> things = new HashMap<>();

    public static Thing get(String variable) {
        return things.get(variable);
    }

    public static void put(String variable, Thing thing) {
        things.put(variable, thing);
    }

    @Then("entity/attribute/relation {var} is null: {bool}")
    public void thing_is_null(String var, boolean isNull) {
        if (isNull) assertNull(get(var));
        else assertNotNull(get(var));
    }

    @Then("entity/attribute/relation {var} is deleted: {bool}")
    public void thing_is_deleted(String var, boolean isDeleted) {
        assertEquals(isDeleted, get(var).isDeleted(tx()));
    }

    @Then("{root_label} {var} has type: {type_label}")
    public void thing_has_type(RootLabel rootLabel, String var, String typeLabel) {
        ThingType type = get_thing_type(rootLabel, typeLabel);
        assertEquals(type, get(var).getType());
    }

    @When("delete entity:/attribute:/relation: {var}")
    public void delete_thing(String var) {
        get(var).delete(tx());
    }

    @When("entity/attribute/relation {var} set has: {var}")
    public void thing_set_has(String var1, String var2) {
        get(var1).setHas(tx(), get(var2).asAttribute());
    }

    @Then("entity/attribute/relation {var} set has: {var}; throws exception")
    public void thing_set_has_throws_exception(String var1, String var2) {
        assertThrows(() -> get(var1).setHas(tx(), get(var2).asAttribute()));
    }

    @When("entity/attribute/relation {var} unset has: {var}")
    public void thing_unset_has(String var1, String var2) {
        get(var1).unsetHas(tx(), get(var2).asAttribute());
    }

    @Then("entity/attribute/relation {var} get keys contain: {var}")
    public void thing_get_keys_contain(String var1, String var2) {
        assertTrue(get(var1).getHas(tx(), set(key())).anyMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get keys do not contain: {var}")
    public void thing_get_keys_do_not_contain(String var1, String var2) {
        assertTrue(get(var1).getHas(tx(), set(key())).noneMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes contain: {var}")
    public void thing_get_attributes_contain(String var1, String var2) {
        assertTrue(get(var1).getHas(tx()).anyMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes\\( ?{type_label} ?) contain: {var}")
    public void thing_get_attributes_contain(String var1, String typeLabel, String var2) {
        assertTrue(get(var1).getHas(tx(), tx().concepts().getAttributeType(typeLabel)).anyMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes\\( ?{type_label} ?) as\\( ?boolean ?) contain: {var}")
    public void thing_get_attributes_as_boolean_contain(String var1, String typeLabel, String var2) {
        assertTrue(get(var1).getHas(tx(), tx().concepts().getAttributeType(typeLabel)).anyMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes\\( ?{type_label} ?) as\\( ?long ?) contain: {var}")
    public void thing_get_attributes_as_long_contain(String var1, String typeLabel, String var2) {
        assertTrue(get(var1).getHas(tx(), tx().concepts().getAttributeType(typeLabel)).anyMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes\\( ?{type_label} ?) as\\( ?double ?) contain: {var}")
    public void thing_get_attributes_as_double_contain(String var1, String typeLabel, String var2) {
        assertTrue(get(var1).getHas(tx(), tx().concepts().getAttributeType(typeLabel)).anyMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes\\( ?{type_label} ?) as\\( ?string ?) contain: {var}")
    public void thing_get_attributes_as_string_contain(String var1, String typeLabel, String var2) {
        assertTrue(get(var1).getHas(tx(), tx().concepts().getAttributeType(typeLabel)).anyMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes\\( ?{type_label} ?) as\\( ?datetime ?) contain: {var}")
    public void thing_get_attributes_as_datetime_contain(String var1, String typeLabel, String var2) {
        assertTrue(get(var1).getHas(tx(), tx().concepts().getAttributeType(typeLabel)).anyMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes do not contain: {var}")
    public void thing_get_attributes_do_not_contain(String var1, String var2) {
        assertTrue(get(var1).getHas(tx()).noneMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes\\( ?{type_label} ?) do not contain: {var}")
    public void thing_get_attributes_do_not_contain(String var1, String typeLabel, String var2) {
        assertTrue(get(var1).getHas(tx(), tx().concepts().getAttributeType(typeLabel)).noneMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes\\( ?{type_label} ?) as\\( ?boolean ?) do not contain: {var}")
    public void thing_get_attributes_as_boolean_do_not_contain(String var1, String typeLabel, String var2) {
        assertTrue(get(var1).getHas(tx(), tx().concepts().getAttributeType(typeLabel)).noneMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes\\( ?{type_label} ?) as\\( ?long ?) do not contain: {var}")
    public void thing_get_attributes_as_long_do_not_contain(String var1, String typeLabel, String var2) {
        assertTrue(get(var1).getHas(tx(), tx().concepts().getAttributeType(typeLabel)).noneMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes\\( ?{type_label} ?) as\\( ?double ?) do not contain: {var}")
    public void thing_get_attributes_as_double_do_not_contain(String var1, String typeLabel, String var2) {
        assertTrue(get(var1).getHas(tx(), tx().concepts().getAttributeType(typeLabel)).noneMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes\\( ?{type_label} ?) as\\( ?string ?) do not contain: {var}")
    public void thing_get_attributes_as_string_do_not_contain(String var1, String typeLabel, String var2) {
        assertTrue(get(var1).getHas(tx(), tx().concepts().getAttributeType(typeLabel)).noneMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get attributes\\( ?{type_label} ?) as\\( ?datetime ?) do not contain: {var}")
    public void thing_get_attributes_as_datetime_do_not_contain(String var1, String typeLabel, String var2) {
        assertTrue(get(var1).getHas(tx(), tx().concepts().getAttributeType(typeLabel)).noneMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get relations\\( ?{scoped_label} ?) contain: {var}")
    public void thing_get_relations_contain(String var1, Label scopedLabel, String var2) {
        assertTrue(get(var1).getRelations(tx(), tx().concepts().getRelationType(scopedLabel.scope().get()).getRelates(tx(), scopedLabel.name())).anyMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get relations contain: {var}")
    public void thing_get_relations_contain(String var1, String var2) {
        assertTrue(get(var1).getRelations(tx()).anyMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get relations\\( ?{scoped_label} ?) do not contain: {var}")
    public void thing_get_relations_do_not_contain(String var1, Label scopedLabel, String var2) {
        assertTrue(get(var1).getRelations(tx(), tx().concepts().getRelationType(scopedLabel.scope().get()).getRelates(tx(), scopedLabel.name())).noneMatch(k -> k.equals(get(var2))));
    }

    @Then("entity/attribute/relation {var} get relations do not contain: {var}")
    public void thing_get_relations_do_not_contain(String var1, String var2) {
        assertTrue(get(var1).getRelations(tx()).noneMatch(k -> k.equals(get(var2))));
    }

    @After
    public void clear() {
        things.clear();
    }
}

File: csharp/Test/Behaviour/Concept/Thing/ThingSteps.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using Xunit;
using Xunit.Gherkin.Quick;

using TypeDB.Driver;
using TypeDB.Driver.Api;
using TypeDB.Driver.Common;
using static TypeDB.Driver.Api.IThingType.Annotation;

namespace TypeDB.Driver.Test.Behaviour
{
    public partial class BehaviourSteps
    {
        private static Dictionary<string, IThing?> _things = new Dictionary<string, IThing?>();

        public static IThing? Get(string variable)
        {
            return _things[variable];
        }

        public static void Put(string variable, IThing? thing)
        {
            _things[variable] = thing;
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) is null: (true|false)")]
        public void ThingIsNull(string rootLabel, string var, bool isNull)
        {
            Assert.Equal(isNull, Get(var) == null);
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) is deleted: (true|false)")]
        public void ThingIsDeleted(string rootLabel, string var, bool isDeleted)
        {
            Assert.Equal(isDeleted, Get(var)!.IsDeleted(Tx).Resolve()!);
        }

        [Then(@"(entity|attribute|relation|thing) \$([a-zA-Z0-9]+) has type: ([a-zA-Z0-9-_]+)")]
        public void ThingHasType(string rootLabel, string var, string typeLabel)
        {
            IThingType type = GetThingType(rootLabel, typeLabel);
            Assert.Equal(type, Get(var)!.Type);
        }

        [When(@"delete (entity|attribute|relation): \$([a-zA-Z0-9]+)")]
        public void DeleteThings(string rootLabel, string var)
        {
            Get(var)!.Delete(Tx).Resolve();
        }

        [When(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) set has: \$([a-zA-Z0-9]+)")]
        public void ThingSetHas(string rootLabel, string var1, string var2)
        {
            Get(var1)!.SetHas(Tx, Get(var2)!.AsAttribute()).Resolve();
        }

        [When(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) set has: \$([a-zA-Z0-9]+); throws exception")]
        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) set has: \$([a-zA-Z0-9]+); throws exception")]
        public void ThingSetHasThrowsException(string rootLabel, string var1, string var2)
        {
            Assert.Throws<TypeDBDriverException>(
                () => Get(var1)!.SetHas(Tx, Get(var2)!.AsAttribute()).Resolve());
        }

        [When(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) unset has: \$([a-zA-Z0-9]+)")]
        public void ThingUnsetHas(string rootLabel, string var1, string var2)
        {
            Get(var1)!.UnsetHas(Tx, Get(var2)!.AsAttribute()).Resolve();
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get keys contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetKeysContain(string rootLabel, string var1, string var2)
        {
            Assert.True(Get(var1)!
                .GetHas(Tx, new []{NewKey()})
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get keys do not contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetKeysDoNotContain(string rootLabel, string var1, string var2)
        {
            Assert.False(Get(var1)!
                .GetHas(Tx, new []{NewKey()})
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesContain(string rootLabel, string var1, string var2)
        {
            Assert.True(Get(var1)!.GetHas(Tx).Where(k => k.Equals(Get(var2)!)).Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes\(([a-zA-Z0-9-_]+)\) contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesContain(string rootLabel, string var1, string typeLabel, string var2)
        {
            Assert.True(Get(var1)!
                .GetHas(Tx, Tx.Concepts.GetAttributeType(typeLabel).Resolve()!)
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes\(([a-zA-Z0-9-_]+)\) as\(boolean\) contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesAsBooleanContain(string rootLabel, string var1, string typeLabel, string var2)
        {
            Assert.True(Get(var1)!
                .GetHas(Tx, Tx.Concepts.GetAttributeType(typeLabel).Resolve()!)
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes\(([a-zA-Z0-9-_]+)\) as\(long\) contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesAsLongContain(string rootLabel, string var1, string typeLabel, string var2)
        {
            Assert.True(Get(var1)!
                .GetHas(Tx, Tx.Concepts.GetAttributeType(typeLabel).Resolve()!)
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes\(([a-zA-Z0-9-_]+)\) as\(double\) contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesAsDoubleContain(string rootLabel, string var1, string typeLabel, string var2)
        {
            Assert.True(Get(var1)!
                .GetHas(Tx, Tx.Concepts.GetAttributeType(typeLabel).Resolve()!)
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes\(([a-zA-Z0-9-_]+)\) as\(string\) contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesAsStringContain(string rootLabel, string var1, string typeLabel, string var2)
        {
            Assert.True(Get(var1)!
                .GetHas(Tx, Tx.Concepts.GetAttributeType(typeLabel).Resolve()!)
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes\(([a-zA-Z0-9-_]+)\) as\(datetime\) contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesAsDatetimeContain(string rootLabel, string var1, string typeLabel, string var2)
        {
            Assert.True(Get(var1)!
                .GetHas(Tx, Tx.Concepts.GetAttributeType(typeLabel).Resolve()!)
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes do not contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesDoNotContain(string rootLabel, string var1, string var2)
        {
            Assert.False(Get(var1)!.GetHas(Tx).Where(k => k.Equals(Get(var2)!)).Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes\(([a-zA-Z0-9-_]+)\) do not contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesDoNotContain(string rootLabel, string var1, string typeLabel, string var2)
        {
            Assert.False(Get(var1)!
                .GetHas(Tx, Tx.Concepts.GetAttributeType(typeLabel).Resolve()!)
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes\(([a-zA-Z0-9-_]+)\) as\(boolean\) do not contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesAsBooleanDoNotContain(string rootLabel, string var1, string typeLabel, string var2)
        {
            Assert.False(Get(var1)!
                .GetHas(Tx, Tx.Concepts.GetAttributeType(typeLabel).Resolve()!)
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes\(([a-zA-Z0-9-_]+)\) as\(long\) do not contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesAsLongDoNotContain(string rootLabel, string var1, string typeLabel, string var2)
        {
            Assert.False(Get(var1)!
                .GetHas(Tx, Tx.Concepts.GetAttributeType(typeLabel).Resolve()!)
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes\(([a-zA-Z0-9-_]+)\) as\(double\) do not contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesAsDoubleDoNotContain(string rootLabel, string var1, string typeLabel, string var2)
        {
            Assert.False(Get(var1)!
                .GetHas(Tx, Tx.Concepts.GetAttributeType(typeLabel).Resolve()!)
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes\(([a-zA-Z0-9-_]+)\) as\(string\) do not contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesAsStringDoNotContain(string rootLabel, string var1, string typeLabel, string var2)
        {
            Assert.False(Get(var1)!
                .GetHas(Tx, Tx.Concepts.GetAttributeType(typeLabel).Resolve()!)
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get attributes\(([a-zA-Z0-9-_]+)\) as\(datetime\) do not contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetAttributesAsDatetimeDoNotContain(string rootLabel, string var1, string typeLabel, string var2)
        {
            Assert.False(Get(var1)!
                .GetHas(Tx, Tx.Concepts.GetAttributeType(typeLabel).Resolve()!)
                .Where(k => k.Equals(Get(var2)!))
                .Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get relations\(([a-zA-Z0-9-_]+:[a-zA-Z0-9-_]+)\) contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetRelationsContain(string rootLabel, string var1, string scopedLabelData, string var2)
        {
            Label scopedLabel = GetScopedLabel(scopedLabelData);

            var relates = Tx.Concepts
                .GetRelationType(scopedLabel.Scope!).Resolve()!
                .GetRelates(Tx, scopedLabel.Name).Resolve()!;

            Assert.True(Get(var1)!.GetRelations(Tx, relates).Where(k => k.Equals(Get(var2)!)).Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get relations contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetRelationsContain(string rootLabel, string var1, string var2)
        {
            Assert.True(Get(var1)!.GetRelations(Tx).Where(k => k.Equals(Get(var2)!)).Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get relations\(([a-zA-Z0-9-_]+:[a-zA-Z0-9-_]+)\) do not contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetRelationsDoNotContain(string rootLabel, string var1, string scopedLabelData, string var2)
        {
            Label scopedLabel = GetScopedLabel(scopedLabelData);

            var relates = Tx.Concepts
                .GetRelationType(scopedLabel.Scope!).Resolve()!
                .GetRelates(Tx, scopedLabel.Name).Resolve()!;

            Assert.False(Get(var1)!.GetRelations(Tx, relates).Where(k => k.Equals(Get(var2)!)).Any());
        }

        [Then(@"(entity|attribute|relation) \$([a-zA-Z0-9]+) get relations do not contain: \$([a-zA-Z0-9]+)")]
        public void ThingGetRelationsDoNotContain(string rootLabel, string var1, string var2)
        {
            Assert.False(Get(var1)!.GetRelations(Tx).Where(k => k.Equals(Get(var2)!)).Any());
        }
    }
}
